src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs

```diff
@@ -207,6 +183,75 @@ int IVsEditorFactory.Close()
             return LoaderName;
         }
 
+        private (string editorCaption, Guid commandUI, IntPtr documentView) GetEditorData(IVsTextBuffer textBuffer, string physicalView, IVsHierarchy vsHierarchy, uint itemId, bool initializeView)
+        {
+            // If the text buffer is marked as read-only, ensure that the padlock icon is displayed
+            // next the new window's title and that [Read Only] is appended to title.
+            var readOnlyStatus = READONLYSTATUS.ROSTATUS_NotReadOnly;
+            if (ErrorHandler.Succeeded(textBuffer.GetStateFlags(out var textBufferFlags)) &&
+                0 != (textBufferFlags & ((uint)BUFFERSTATEFLAGS.BSF_FILESYS_READONLY | (uint)BUFFERSTATEFLAGS.BSF_USER_READONLY)))
+            {
+                readOnlyStatus = READONLYSTATUS.ROSTATUS_ReadOnly;
+            }
+
+            return physicalView switch
+            {
+                FormPhysicalViewName => GetWinFormsDesignerData(vsHierarchy, itemId, textBuffer, readOnlyStatus, initializeView),
+                CodePhysicalViewName => GetCodeEditorData(textBuffer, readOnlyStatus, initializeView),
+                _ =>  (string.Empty, Guid.Empty, IntPtr.Zero),
+            };
+        }
+
+        private (string editorCaption, Guid commandUI, IntPtr documentView) GetWinFormsDesignerData(IVsHierarchy vsHierarchy, uint itemid, IVsTextBuffer textBuffer, READONLYSTATUS readOnlyStatus, bool initializeView)
+        {
+            var loaderName = GetWinFormsLoaderName(vsHierarchy);
+            var designerService = (IVSMDDesignerService)_oleServiceProvider.QueryService<SVSMDDesignerService>();
+            var designerLoader = (IVSMDDesignerLoader)designerService.CreateDesignerLoader(loaderName);
+            if (designerLoader is null)
+            {
+                return GetCodeEditorData(textBuffer, readOnlyStatus, initializeView);
+            }
+
+            try
+            {
+                if (initializeView)
```

Comment:
This is the part where I am not sure if we can optimize the code more by not initializing stuff when we don't create the view. It may be OK to short-circuit the whole thing and return the known value.

Reply:
Since the view is derived from the designer, which can only be created after the designer loader is initialized, I'm not sure how any of this can be deferred.

Reply:
Yes, that is my concern, We have 2 options to improve it -
1. Change the code in the designer and loader to return the value if possible, without doing much work
2. Short-circuit the code here

My thinking is that keeping things such that they look like regular execution over here is preferable. We can make the required changes in the loader to get the info cheaply.

What do you think?

Reply:
I'm actually pretty concerned about what might happen with the subsequent code if `IVSMDDesignerLoader.Initialize(...)` is not called. That method sets up a ton of services that the other calls here likely depend on ([code](http://index/?leftProject=Microsoft.VisualStudio.Design&leftSymbol=c7eyce2zor1k&file=Serialization%5CCodeDom%5CVSCodeDomDesignerLoader.cs&line=556)). Calling `GetEditorCaption` looks relatively safe, assuming that it never changes ([code](http://index/?leftProject=Microsoft.VisualStudio.Design&leftSymbol=t4z7ygts5tsv&file=Serialization%5CCodeDom%5CVSCodeDomDesignerLoader.cs&line=1733)). However, there's no telling what `CreateDesigner(...)` might do ([code](http://index/?leftProject=Microsoft.VisualStudio.Design&leftSymbol=n8p1tszkfyz7&file=DesignerActivationService.cs&line=629)). This makes me pretty nervous. Changing the old classic designer and new designer loader is something I would be very cautious of.

Looping in @dreddy-work from the WinForms team, who was the last person to change the designer activation logic.

Reply:
Yes, that is my concern too. If there are no objections, I want to put an empty implementation here and keep doing what we do in VS to special-case WinForms designer for this case. The requirement for me is to get the docdata which looks OK from the implementation here. The WinForms team has the most context and they can update these methods later to keep things working safely.

Reply:
I vaguely remember order of initialize -> create designer was coupled. I will have to go over that code one more time 
to recollect. @Mohit-Chakraborty , did you try a scenario with Netcore and classic winforms projects in a single solution and try opening the designers with these changes?

Reply:
I reverted most of the changes that impact the existing API.